[PyTorch] Outline OperatorEntry::assertSignatureIsCorrect fail path

swolchok opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #51368 [PyTorch] Use real if constexpr behind macro in hot template
* #51367 [PyTorch] Debug-gate static_assert in KernelFunction::makeFromUnboxedFunctor
* #51319 [PyTorch] Remove reference_cast in make_boxed_from_unboxed_functor
* #51463 [PyTorch] Fix return value of IValue::to for Tensor/String
* #51315 [PyTorch][easy] Fix TODOs in CppFunction constructors
* #51313 [PyTorch] Reduce template expansion in call_functor_with_args_from_stack
* #51307 [PyTorch] Remove always-empty string args to inferFunctionSchemaFromFunctor
* #51293 [PyTorch] IWYU in torch/csrc/utils/future.h
* **#51269 [PyTorch] Outline OperatorEntry::assertSignatureIsCorrect fail path**
* #51247 [PyTorch] check isValidUnboxed() in the dispatcher
* #51165 [PyTorch] Pass TensorOptions by value
* #51245 [PyTorch] Add C10_ALWAYS_INLINE to critical dispatcher paths
* #51163 [PyTorch] Refactor Dispatcher to inline less code in fast path

This saves about 10% of the compile time of Functions.cpp. Found using clang-9's `-ftime-trace` feature + ClangBuildAnalyzer.

Differential Revision: [D26121814](https://our.internmc.facebook.com/intern/diff/D26121814/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D26121814/)!

swolchok committed:
[PyTorch] Outline OperatorEntry::assertSignatureIsCorrect fail path

This saves about 10% of the compile time of Functions.cpp. Found using clang-9's `-ftime-trace` feature + ClangBuildAnalyzer.

Differential Revision: [D26121814](https://our.internmc.facebook.com/intern/diff/D26121814/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D26121814/)!

[ghstack-poisoned]

swolchok committed:
fix build on "[PyTorch] Outline OperatorEntry::assertSignatureIsCorrect fail path"


This saves about 10% of the compile time of Functions.cpp. Found using clang-9's `-ftime-trace` feature + ClangBuildAnalyzer.

Differential Revision: [D26121814](https://our.internmc.facebook.com/intern/diff/D26121814/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D26121814/)!

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] Outline OperatorEntry::assertSignatureIsCorrect fail path"


This saves about 10% of the compile time of Functions.cpp. Found using clang-9's `-ftime-trace` feature + ClangBuildAnalyzer.

Differential Revision: [D26121814](https://our.internmc.facebook.com/intern/diff/D26121814/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D26121814/)!

[ghstack-poisoned]

ezyang commented on aten/src/ATen/core/dispatch/OperatorEntry.cpp:
```diff
@@ -362,6 +362,19 @@ std::string OperatorEntry::listAllDispatchKeys() const {
   return str.str();
 }
 
+void OperatorEntry::reportSignatureError(std::string name) const {
```
`std::string` by value? (eyebrow quirk)

swolchok replied:
It's coming from a prvalue so there's no point in trying to pass that by reference -- https://www.internalfb.com/intern/diffusion/FBS/browsefile/master/fbcode/caffe2/aten/src/ATen/core/dispatch/CppSignature.h?commit=5d6752aa058ea0b4b2241ff90efc86d9d38c06d3&lines=28-30

See https://godbolt.org/z/6r6c38

swolchok committed:
Update on "[PyTorch] Outline OperatorEntry::assertSignatureIsCorrect fail path"


This saves about 10% of the compile time of Functions.cpp. Found using clang-9's `-ftime-trace` feature + ClangBuildAnalyzer.

Differential Revision: [D26121814](https://our.internmc.facebook.com/intern/diff/D26121814/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D26121814/)!

[ghstack-poisoned]
